plugins/gpio/fu-plugin-gpio.c

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (C) 2022 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <fwupdplugin.h>
+
+#include "fu-gpio-device.h"
+
+struct FuPluginData {
+	FuGpioDevice *gpio_device;
```

Comment:
Is this a valid assumption that there is only ever 1 GPIO device on a system?

Reply:
That's an excellent question. There can be multiple GPIO devices, but we can only control one pin of one of them for the prepare->cleanup action. I think renaming the variable to `gpio_device_active` might make that more obvious. A better idea still might just be to look up the device from the plugin cache from the logical_id, which means less ref's everywhere, and also means if the gpio device someone weirdly disappears during the update then the right thing happens -- and we also solve the race issue you point out.

Reply:
Those changes make it a lot clearer what is going on.